File: cdm-sources/cdm/product/qualification/functions/Qualify_BaseProduct_EquitySwap.java
package cdm.product.qualification.functions;

import cdm.product.template.EconomicTerms;
import cdm.product.template.Payout;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.expression.CardinalityOperator;
import com.rosetta.model.lib.expression.ComparisonResult;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperS;
import javax.inject.Inject;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(Qualify_BaseProduct_EquitySwap.Qualify_BaseProduct_EquitySwapDefault.class)
public abstract class Qualify_BaseProduct_EquitySwap implements RosettaFunction {
	
	// RosettaFunction dependencies
	//
	@Inject protected PerformancePayoutAndFixedPricePayoutOnlyExists performancePayoutAndFixedPricePayoutOnlyExists;
	@Inject protected PerformancePayoutAndInterestRatePayoutOnlyExists performancePayoutAndInterestRatePayoutOnlyExists;
	@Inject protected PerformancePayoutOnlyExists performancePayoutOnlyExists;
	@Inject protected Qualify_AssetClass_Equity qualify_AssetClass_Equity;

	/**
	* @param economicTerms 
	* @return is_product 
	*/
	public Boolean evaluate(EconomicTerms economicTerms) {
		Boolean is_product = doEvaluate(economicTerms);
		
		return is_product;
	}

	protected abstract Boolean doEvaluate(EconomicTerms economicTerms);

	public static class Qualify_BaseProduct_EquitySwapDefault extends Qualify_BaseProduct_EquitySwap {
		@Override
		protected Boolean doEvaluate(EconomicTerms economicTerms) {
			Boolean is_product = null;
			return assignOutput(is_product, economicTerms);
		}
		
		protected Boolean assignOutput(Boolean is_product, EconomicTerms economicTerms) {
			is_product = areEqual(MapperS.of(qualify_AssetClass_Equity.evaluate(economicTerms)), MapperS.of(true), CardinalityOperator.All).and(ComparisonResult.of(MapperS.of(performancePayoutAndInterestRatePayoutOnlyExists.evaluate(MapperS.of(economicTerms).<Payout>mapC("getPayout", _economicTerms -> _economicTerms.getPayout()).getMulti()))).or(ComparisonResult.of(MapperS.of(performancePayoutAndFixedPricePayoutOnlyExists.evaluate(MapperS.of(economicTerms).<Payout>mapC("getPayout", _economicTerms -> _economicTerms.getPayout()).getMulti())))).or(ComparisonResult.of(MapperS.of(performancePayoutOnlyExists.evaluate(MapperS.of(economicTerms).<Payout>mapC("getPayout", _economicTerms -> _economicTerms.getPayout()).getMulti()))))).get();
			
			return is_product;
		}
	}
}

File: cdm-sources/src/cdm/product/qualification/functions/Qualify_BaseProduct_EquitySwap.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.product.qualification.functions.PerformancePayoutOnlyExists import PerformancePayoutOnlyExists
from cdm.product.qualification.functions.PerformancePayoutAndInterestRatePayoutOnlyExists import PerformancePayoutAndInterestRatePayoutOnlyExists
from cdm.product.template.EconomicTerms import EconomicTerms
from cdm.product.qualification.functions.PerformancePayoutAndFixedPricePayoutOnlyExists import PerformancePayoutAndFixedPricePayoutOnlyExists

__all__ = ['Qualify_BaseProduct_EquitySwap']


@replaceable
def Qualify_BaseProduct_EquitySwap(economicTerms: EconomicTerms) -> bool:
    """
    Qualifies a product as having the Asset Class classification Equity and Base Product Classification Swap.
    
    Parameters 
    ----------
    economicTerms : EconomicTerms
    
    Returns
    -------
    is_product : boolean
    
    """
    self = inspect.currentframe()
    
    
    is_product =  (all_elements(Qualify_AssetClass_Equity(rosetta_resolve_attr(self, "economicTerms")), "=", True) and ((PerformancePayoutAndInterestRatePayoutOnlyExists(rosetta_resolve_attr(rosetta_resolve_attr(self, "economicTerms"), "payout")) or PerformancePayoutAndFixedPricePayoutOnlyExists(rosetta_resolve_attr(rosetta_resolve_attr(self, "economicTerms"), "payout"))) or PerformancePayoutOnlyExists(rosetta_resolve_attr(rosetta_resolve_attr(self, "economicTerms"), "payout"))))
    
    
    return is_product

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)
